var/spack/repos/builtin/packages/kokkos/package.py

```diff
@@ -345,12 +354,26 @@ def cmake_args(self):
             if spec.variants[tpl].value:
                 options.append(self.define(tpl + "_DIR", spec[tpl].prefix))
 
-        if "+rocm" in self.spec:
-            options.append(self.define("CMAKE_CXX_COMPILER", self.spec["hip"].hipcc))
-        elif "+wrapper" in self.spec:
+        if "+wrapper" in self.spec:
             options.append(
                 self.define("CMAKE_CXX_COMPILER", self.spec["kokkos-nvcc-wrapper"].kokkos_cxx)
             )
+        elif "+rocm" in self.spec:
+            if "+compile_as_cmake_language" in self.spec:
+                options.append(
+                    self.define(
+                        "CMAKE_HIP_COMPILER",
+                        join_path(self.spec["llvm-amdgpu"].prefix.bin, "amdclang++"),
+                    )
+                )
+                options.append(from_variant("CMAKE_HIP_STANDARD", "cxxstd"))
+            else:
+                options.append(self.define("CMAKE_CXX_COMPILER", self.spec["hip"].hipcc))
+        elif "+cuda" in self.spec and "+compile_as_cmake_language" in self.spec:
+            options.append(
+                self.define("CMAKE_CUDA_COMPILER", join_path(self.spec["cuda"].prefix.bin, "nvcc"))
```

Comment:
Same here. Feel strange to hard code compiler name. I am not sure that it always works for external cuda install.

Reply:
As I see the message "This comment was marked as resolved." and the comment by @cedricchevalier19, I mark this conversation as resolved.